### ISSUE TITLE
Could not open configuration file /etc/httpd/${APP_HOME_DIR}/vhost.conf

### ISSUE THREAD

**Issue by MatthewHerbst** [jdeathe/centos-ssh-apache-php#90]
So, I created my own image using this one as a base.

My Dockerfile:

```
FROM jdeathe/centos-ssh-apache-php:latest
MAINTAINER xxx <xxx.yyy>

# -----------------------------------------------------------------------------
# Install Git, Node.js/npm, php-mysql
# -----------------------------------------------------------------------------
RUN curl --silent --location https://rpm.nodesource.com/setup_4.x | bash -
RUN rpm --rebuilddb\
        && yum --setopt=tsflags=nodocs -y install \
        git \
        nodejs \
        php-mysql \
        && yum versionlock add \
        git \
        nodejs \
        php-mysql \
        && rm -rf /var/cache/yum/* \
        && yum clean all

# -----------------------------------------------------------------------------
# Populate the app home directory
# -----------------------------------------------------------------------------
ADD package.json /var/www/app/
ADD gulpfile.js /var/www/app/
ADD src /var/www/app/src/

# -----------------------------------------------------------------------------
# Build the app
# -----------------------------------------------------------------------------
RUN npm --prefix /var/www/app/ install /var/www/app/ \
 && npm --prefix /var/www/app/ run prod \
 && cp -R /var/www/app/dist/* /var/www/app/public_html/

EXPOSE 80 8443 443
```

This builds fine, but when I go to the address of my container in my browser it doesn't connect (but just running this base image works fine). So, to investigate I went onto the host using `docker run -i -t xxxx /bin/bash` and found this:

```
[root@b93d38591ec0 /]# service httpd restart
Stopping httpd:                                            [FAILED]
Starting httpd: httpd: Syntax error on line 1019 of /etc/httpd/conf/httpd.conf: Could not open configuration file /etc/httpd/${APP_HOME_DIR}/vhost.conf: No such file or directory
                                                           [FAILED]
[root@b93d38591ec0 /]# echo $APP_HOME_DIR
/var/www/app
[root@b93d38591ec0 /]# ls /etc/httpd/                          
conf  conf.d  logs  modules  run
```

I also tried adding in all the ENV declarations from this package into my Dockerfile but that had no effect. What's might be causing this issue? Clearly I'm missing something since the base image works fine.

**Comment by MatthewHerbst** [comment]
The error was due to be not using the `-v` flag. It's not clear to me why I need a data volume if I don't need the container to persist data. Can this be disabled?

**Comment by jdeathe** [comment]
A data volume is not a requirement of the source image.

I would use the following steps to debug.

1. The environment variable for `APP_HOME_DIR` has been set as expected using.

```
$ docker exec -it {container-name} env | grep APP_HOME_DIR
```

2. Check that the httpd configuration file `/etc/httpd/conf/httpd.conf` has the include path set correctly. Ref: https://github.com/jdeathe/centos-ssh-apache-php/blob/master/etc/services-config/httpd/conf/httpd.conf#L1019

The fact the error log reports `/etc/httpd/${APP_HOME_DIR}/vhost.conf` suggests that the environment variable `APP_HOME_DIR}` has not been set in the running container so the configuration script is attempting to include a file path relative to itself. 

```
$ docker exec -it {container-name} tail /etc/httpd/conf/httpd.conf
```

My guess is that you have a syntax error in your run command that is fixed by adding in the `-v` part.

**Comment by MatthewHerbst** [comment]
This is the run command I am using without the `-v`:
```
docker run -d --name my-cool-app -p 8080:80 my-cool-container
```
Wouldn't the command fail to run with a syntax error anyways?

The env variable seems set fine:
```
$ docker exec -it {container-name} env | grep APP_HOME_DIR
APP_HOME_DIR=/var/www/app
```

I even tried declaring
`ENV APP_HOME_DIR /var/www/app`
in my Dockerfile right below the ENV call I already have, and that had no effect.

**Comment by jdeathe** [comment]
@MatthewHerbst 

I would stop and delete the container with the issue. Then rebuild the image and run it again.

**Comment by MatthewHerbst** [comment]
So, after further investigation:

The image works fine with my image, I just had to change

```
RUN rpm --rebuilddb \
        && yum --setopt=tsflags=nodocs -y install \
        php-mysql
        && yum versionlock add \
        php-mysql \
        && rm -rf /var/cache/yum/* \
        && yum clean all
```

to

```
RUN rpm --rebuilddb \
        && yum --setopt=tsflags=nodocs -y install \
        php-pdo-5.3.3-46.el6_6.x86_64 \
        php-mysql-5.3.3-46.el6_6.x86_64 \
        && yum versionlock add \
        php-pdo \
        php-mysql \
        && rm -rf /var/cache/yum/* \
        && yum clean all
```

Without specifying the `php-mysql`/`php-pdo` versions (*and specifically with `.x86_64`) at the end, the  build actually was throwing an error that was just getting lost in the build log. Making the change above causes the site to work (without or without the `-v` option)!

Now, as to the original error of this thread, I'm still not sure what the issue is. I agree it's likely a relativity issue. However, other than being a little misleading when going bash on the box, it isn't stopping the site from running.

@jdeathe should this still be an issue? The behavior with `service httpd restart` is indeed strange, but it doesn't stop the site from working. It's actually an issue that can be found just by using this project:

```
bash-3.2$ docker run -i -t jdeathe/centos-ssh-apache-php:centos-6 /bin/bash 
[root@722ebed09fd5 /]# service httpd restart
Stopping httpd:                                            [FAILED]
Starting httpd: httpd: Syntax error on line 1019 of /etc/httpd/conf/httpd.conf: Could not open configuration file /etc/httpd/${APP_HOME_DIR}/vhost.conf: No such file or directory
                                                           [FAILED]
[root@722ebed09fd5 /]# 
```

**Comment by jdeathe** [comment]
I have made improvements to the build.sh helper script in the source image
(jdeathe/centos-ssh) for that same reason; build errors could easily go
unnoticed. I will get the build script updated on this image too in the
next release or at least  by 1.5.0.

Using the /etc/init.d/httpd directly to restart the service isn't advisable
and might be what's causing misleading behaviour. The container uses
Supervisor to run the httpd service in the foreground and the init script
is unused. That being said, I have used apachectl commands from the
container without issue before.

**Comment by MatthewHerbst** [comment]
@jdeathe I just went back into the container based on your last comment, and yeah, the `apachectl` commands work fine (other than a complaint about not being able to determine the full-qualified domain name).

In the event that I needed to restart httpd for a change to take effect, what is the recommended way to have Supervisor restart httpd?

```
[root@9de6944d71d3 /]# service supervisor
supervisor: unrecognized service
[root@9de6944d71d3 /]# supervisorctl
Error: .ini file does not include supervisorctl section
For help, use /usr/bin/supervisorctl -h
```

Looking forward to 1.5.0!
